samples/unity-of-bugs/Assets/Scripts/SmokeTester.cs

```diff
@@ -75,19 +75,62 @@ public void Start()
 
     public static void SmokeTest()
     {
+        var exitCode = 0;
+        void Exit(int code)
+        {
+            if (exitCode != 0)
+            {
+                Debug.Log($"Ignoring spurious Exit({code}). Application is already exiting with code {exitCode}");
+            }
+            else
+            {
+                exitCode = code;
+                Application.Quit(code);
+                // Application.Quit doesn't actually terminate immediately so exit the context at least...
+                throw new Exception($"Quitting with exit code {code}");
+            }
+        }
+
         try
         {
             Debug.Log("SMOKE TEST: Start");
-            var evt = new ManualResetEventSlim();
+            var evt = new AutoResetEvent(false);
 
             var requests = new List<string>();
             void Verify(HttpRequestMessage message)
             {
-                Debug.Log("SMOKE TEST: Verify invoked.");
-                requests.Add(message.Content.ReadAsStringAsync().Result);
+                var msgText = message.Content.ReadAsStringAsync().Result;
+                Debug.Log($"SMOKE TEST: Intercepted HTTP Request #{requests.Count} = {msgText}");
+                requests.Add(msgText);
                 evt.Set();
             }
 
+            var testNumber = 0;
+            void Expect(String message, bool result)
+            {
+                testNumber++;
+                Debug.Log($"SMOKE TEST | {testNumber}. {message}: {(result ? "PASS" : "FAIL")}");
+                if (!result)
+                {
+                    Debug.Log($"SMOKE TEST: Quitting due to a failed test case #{testNumber}");
+                    Exit(testNumber);
+                }
+            }
+
+            void ExpectMessage(int index, String substring)
+            {
+                Debug.Log($"SMOKE TEST: Checking if the HTTP request #{index} contains: '{substring}'.");
+                while (requests.Count < index + 1)
```

Comment:
This will end up throwing, we need to swap the `List` with a `ConcurrentBag`

Reply:
yes, this wasn't finished yet. I'm now locking manually. `ConcurrentBag` wouldn't do because the checks rely on index-based access for messages. Alternatively, I could use a `SyncrhonizedDictionary` with an atomic increment but don't see much benefit in that, at least not code-wise. And I don't know enough about the available c# synchronization primitives & mechanisms to tell whether it would make any difference.